plugin/dnstap/msg/msg.go

```diff
@@ -1,159 +1,78 @@
 package msg
 
 import (
-	"errors"
+	"fmt"
 	"net"
-	"strconv"
 	"time"
 
 	tap "github.com/dnstap/golang-dnstap"
-	"github.com/miekg/dns"
 )
 
-// Builder helps to build a Dnstap message.
-type Builder struct {
-	Packed      []byte
-	SocketProto tap.SocketProtocol
-	SocketFam   tap.SocketFamily
-	Address     net.IP
-	Port        uint32
-	TimeSec     uint64
-	TimeNsec    uint32
-
-	err error
-}
-
-// New returns a new Builder
-func New() *Builder {
-	return &Builder{}
-}
-
-// Addr adds the remote address to the message.
-func (b *Builder) Addr(remote net.Addr) *Builder {
-	if b.err != nil {
-		return b
-	}
-
-	switch addr := remote.(type) {
+// SetQueryAddress adds the query address to the message. This also sets the SocketFamily and SocketProtocol.
+func SetQueryAddress(t *tap.Message, addr net.Addr) error {
+	inet := tap.SocketFamily_INET
+	t.SocketFamily = &inet
+	switch a := addr.(type) {
 	case *net.TCPAddr:
-		b.Address = addr.IP
-		b.Port = uint32(addr.Port)
-		b.SocketProto = tap.SocketProtocol_TCP
+		t.QueryAddress = a.IP
+		p := uint32(a.Port)
+		t.QueryPort = &p
+		p1 := tap.SocketProtocol_TCP
+		t.SocketProtocol = &p1
+		if a.IP.To4() == nil {
+			inet6 := tap.SocketFamily_INET6
+			t.SocketFamily = &inet6
+		}
+		return nil
 	case *net.UDPAddr:
-		b.Address = addr.IP
-		b.Port = uint32(addr.Port)
-		b.SocketProto = tap.SocketProtocol_UDP
+		t.QueryAddress = a.IP
+		p := uint32(a.Port)
+		t.QueryPort = &p
+		p1 := tap.SocketProtocol_UDP
+		t.SocketProtocol = &p1
+		if a.IP.To4() == nil {
+			inet6 := tap.SocketFamily_INET6
+			t.SocketFamily = &inet6
+		}
+		return nil
 	default:
-		b.err = errors.New("unknown remote address type")
-		return b
-	}
-
-	if b.Address.To4() != nil {
-		b.SocketFam = tap.SocketFamily_INET
-	} else {
-		b.SocketFam = tap.SocketFamily_INET6
+		return fmt.Errorf("unknown address type: %T", a)
 	}
-	return b
 }
 
-// Msg adds the raw DNS message to the dnstap message.
-func (b *Builder) Msg(m *dns.Msg) *Builder {
-	if b.err != nil {
-		return b
-	}
-
-	b.Packed, b.err = m.Pack()
-	return b
-}
-
-// HostPort adds the remote address as encoded by dnsutil.ParseHostPortOrFile to the message.
-func (b *Builder) HostPort(addr string) *Builder {
-	ip, port, err := net.SplitHostPort(addr)
-	if err != nil {
-		b.err = err
-		return b
-	}
-	p, err := strconv.ParseUint(port, 10, 32)
-	if err != nil {
-		b.err = err
-		return b
-	}
-	b.Port = uint32(p)
-
-	if ip := net.ParseIP(ip); ip != nil {
-		b.Address = []byte(ip)
-		if ip := ip.To4(); ip != nil {
-			b.SocketFam = tap.SocketFamily_INET
-		} else {
-			b.SocketFam = tap.SocketFamily_INET6
-		}
-		return b
+// SetResponseAddress the response address to the message.
+func SetResponseAddress(t *tap.Message, addr net.Addr) error {
+	switch a := addr.(type) {
+	case *net.TCPAddr:
+		t.ResponseAddress = a.IP
+		p := uint32(a.Port)
+		t.ResponsePort = &p
+		return nil
```

Comment:
I believe the proto and family should be set as well, since `SetQueryAddress()` and `SetResponseAddress()` can be called independently.

Reply:
done